app/src/animation/animation.c

```diff
@@ -0,0 +1,167 @@
+
+#define PHANDLE_TO_PIXEL(idx, node_id, prop)                                                       \
+    {                                                                                              \
+        .id = idx,                                                                                 \
+        .position_x = DT_PHA_BY_IDX(node_id, prop, idx, position_x),                               \
+        .position_y = DT_PHA_BY_IDX(node_id, prop, idx, position_y),                               \
+        .animation = DEVICE_DT_GET(DT_PHANDLE_BY_IDX(node_id, prop, idx)),                         \
+    },
+
+/**
+ * LED Driver device pointers.
+ */
+static const struct device *drivers[] = {
+    ZMK_DT_INST_FOREACH_PROP_ELEM(0, drivers, PHANDLE_TO_DEVICE)};
+
+/**
+ * Size of the LED driver device pointers array.
+ */
+static const size_t drivers_size = DT_INST_PROP_LEN(0, drivers);
+
+/**
+ * Array containing the number of LEDs handled by each device.
+ */
+static const uint8_t pixels_per_driver[] = {
+    ZMK_DT_INST_FOREACH_PROP_ELEM(0, drivers, PHANDLE_TO_CHAIN_LENGTH)};
+
+/**
+ * Pointers to all active animation devices.
+ */
+static const struct device *animations[] = {
+    ZMK_DT_INST_FOREACH_PROP_ELEM(0, animations, PHANDLE_TO_DEVICE)};
+
+/**
+ * Size of the animation device pointers array.
+ */
+static const size_t animations_size = DT_INST_PROP_LEN(0, animations);
+
+/**
+ * Pixel configuration.
+ */
+static const struct animation_pixel pixels[] = {
+    ZMK_DT_INST_FOREACH_PROP_ELEM(0, pixels, PHANDLE_TO_PIXEL)};
+
+/**
+ * Size of the pixels array.
+ */
+static const size_t pixels_size = DT_INST_PROP_LEN(0, pixels);
+
+/**
+ * Buffer for RGB values ready to be sent to the drivers.
+ */
+static struct led_rgb px_buffer[DT_INST_PROP_LEN(0, pixels)];
+
+/**
+ * Conditional implementation of zmk_animation_get_pixel_by_key_position
+ * if key-pixels is set.
+ */
+#if DT_INST_NODE_HAS_PROP(0, key_position)
+static const uint8_t pixels_by_key_position[] = DT_INST_PROP(0, key_pixels);
+
+size_t zmk_animation_get_pixel_by_key_position(size_t key_position) {
+    return pixels_by_key_position[key_position];
+}
+#endif
+
+#if defined(CONFIG_ZMK_ANIMATION_PIXEL_DISTANCE) && (CONFIG_ZMK_ANIMATION_PIXEL_DISTANCE == 1)
+
+/**
+ * Lookup table for distance between any two pixels.
+ */
+static uint16_t pixel_distance[DT_INST_PROP_LEN(0, pixels)][DT_INST_PROP_LEN(0, pixels)];
+
+uint16_t zmk_animation_get_pixel_distance(size_t pixel_idx, size_t other_pixel_idx) {
+    return pixel_distance[pixel_idx][other_pixel_idx];
+}
+
+#endif
+
+static void zmk_animation_tick(struct k_work *work) {
+    for (size_t i = 0; i < animations_size; ++i) {
+        animation_on_before_frame(animations[i]);
+    }
+
+    for (size_t i = 0; i < pixels_size; ++i) {
+        struct zmk_color_rgb rgb = {
+            .r = 0,
+            .g = 0,
+            .b = 0,
+        };
+
+        animation_render_pixel(pixels[i].animation, &pixels[i], &rgb);
```

Comment:
That's kind of the point with the other methods of the API. If an animation doesn't need to calculate anything for every pixel individually, it should just use `on_before_frame` or `on_after_frame` to update its state and then it just returns those values for each pixel.
This also allows us not to make any assumptions about what pixels are 'grouped' together or not. It should just work.

Reply:
I'm not sure I understand. This appears to unconditionally call animation_render_pixel() for every pixel on ever tick.

Maybe this isn't an issue with ARM processors, but my concern was that running a function 104 times to update every LED on a full-sized board might be spending a lot of time simply making function calls compared to running one function that sets 104 values.

Reply:
I see what you mean. Yeah, technically we could pass the whole buffer but then also, like you already said, animations must be aware of what pixels they're allowed to touch.

This creates an issue in that if you're not able to assign an animation per pixel, we'll end up with a list of animations in which every animation would be applied to all the pixels.
That's where the mask comes in handy but there's a catch. Especially with composition, you could have your regular backlight animation and in case of caps-lock, combine it with a caps-lock indicator animation.  
This in turn, would cause the backlight animation to be called twice - once by the animation system directly, and then a second time by the composition animation.

To solve this, each animation would need to have its own pixel buffer. Which would presumably be only written to 'once' per frame and any subsequent calls just return the same values.
The configs for animations would grow as they'd now need to know which pixels they're responsible for, but I think we should still keep a global reference of pixels' physical position or the keys they're placed under.

I'll play with it a little, see what comes out of it.

On a side note, I admit I haven't looked at the performance yet, I'm probably also not the first person to ask when it comes to that. But ZMK being built on top of Zephyr, ARM CPUs are pretty much as low as it gets?

Reply:
> This creates an issue in that if you're not able to assign an animation per pixel, we'll end up with a list of animations in which every animation would be applied to all the pixels.
That's where the mask comes in handy but there's a catch. Especially with composition, you could have your regular backlight animation and in case of caps-lock, combine it with a caps-lock indicator animation.
This in turn, would cause the backlight animation to be called twice - once by the animation system directly, and then a second time by the composition animation.

What I was thinking is that you'd have the animation system call only the top-level animation, which would then call any child animations, and so on. That would result in some level of duplicate writes though, which isn't ideal, but the same animation would never be called twice.

For example, if we were to organize it in the way I'd suggested, we would have a caps lock animation with backlight animation as its child. On a tick of the caps lock animation, it would run call the backlight animation tick, which would update every pixel, and then the caps lock animation would overwrite one pixel from that as needed. (This would all be done to the buffer of colors, and we would only update the LED driver once at the end.)

Alternatively, if you had composition with animations per pixel, you could somehow track what animation is "under" each one so each animation can defer to the next animation as needed. I'm not sure how you'd do that efficiently though.

Reply:
So what you're saying, the animation relationships would strictly have to be a tree and not a graph? As in, any animation can have a single 'parent' at most?  

Depending on what your setup is, then even something like a caps lock indicator animation that needs to go on top of everything can cause a lot of duplication as you'd now have to duplicate all other key animations.  
This also has the potential of some of these 'duplicates' to run out of sync, or in case of effect animations, they might not affect certain pixels if a duplicate is used.  
I think it can be worked around but I'd prefer to avoid it altogether if I can.

I wonder... if we defined individual animation drivers as children on the main `animation` node instead of `/`., we could get a list of references to all active animations using `DT_FOREACH_CHILD`.  
Then do two function calls:

- `render_frame`: Animations will render the next frame into their internal buffer. This would be run on every animation directly on each re-render. If an animation isn't active, it can choose to do nothing.
- `get_pixels`: Copies values from the internal pixel buffer into the main pixel buffer. In this case we'd only call this on the 'top level' animation(s) and do the chaining from there.

I believe the internal buffers could only be the size of the pixels assigned to any given animation, so that saves some space too. Composition animations would likely be an edge case here, where they'd do majority of their work during `get_pixels` and not `render_frame`.  
I guess either way, this should save us some function calls compared to doing per-pixel calculation, also with composition where the way I have it currently, it'd make another call for every pixel for every composed animation as well.

Reply:
> So what you're saying, the animation relationships would strictly have to be a tree and not a graph? As in, any animation can have a single 'parent' at most?

Correct. I think this makes things a lot simpler, though not necessarily *better*.

> Depending on what your setup is, then even something like a caps lock indicator animation that needs to go on top of everything can cause a lot of duplication as you'd now have to duplicate all other key animations.

I'm not sure what you mean by this. If for example you have several animations to select between, but all use the same caps indicator, then your tree would look like

```
+ caps
  + animation selector
    + animation 1
    + animation 2
    + ...
```

If you didn't want it to be the same for every animation, then it would require duplicates, but I'm also not sure what you mean by

> This also has the potential of some of these 'duplicates' to run out of sync, or in case of effect animations, they might not affect certain pixels if a duplicate is used.

Do you mean synchronization between halves of a split, or synchronization between different running animations on the same half? If the latter, then I was thinking that wouldn't be a problem because each animation would only do it's per-frame processing inside its "tick" function, which would only be called if the animation was active, and each animation could reset its state upon being activated.

For example, consider a setup where you can choose between a solid background or a pair of animations that each cover half the keyboard:

```
+ caps
  + animation selector
    + solid color
    + composite
      + rgb wave 1
      + rgb wave 2
```

On system init/animation activate we would do something like (pseudocode)

```c
// system init
animation_activate(root_animation); // activates caps

// caps activate()
register_caps_state_listener(&data->callback, zmk_animation_request_tick);
animation_activate(data->child); // activates animation selector

// animation selector activate()
data->selected_anim = load_selected_animation();
data->selected_anim = MIN(data->selected_anim, data->children_size - 1);
animation_activate(data->children[data->selected_anim]); // activates either solid color or composite

// solid color activate()
// do nothing

// composite activate()
for (i = 0; i < data->children_size; i++) {
  animation_activate(data->children[i]); // activates both RGB wave instances
}

// rgb wave activate()
data->time_elapsed = 0;
zmk_animation_timer_ref();
```

And animation tick

```c
// zmk_animation_tick()
uint32_t now = k_uptime_get();
uint32 time_elapsed = now - last_tick_time;
last_tick_time = now;

// just pass each animation the same pixel buffer? Each animation 
// needs to know which pixels it should affect, but this avoids
// needing to composite animation-specific buffers together later.
animation_tick(root_animation, time_elapsed, pixels, pixels_size);
update_leds(pixels);

// caps tick()
animation_tick(data->child, time_elasped, pixels, pixels_size);
if (is_caps_on()) {
  pixels[data->caps_pixel] = data->caps_color;
}

// animation selector tick()
animation_tick(data->children[data->selected_anim], elapsed_time, pixels, pixels_size);

// solid color tick()
// Ideally would have a way to use a subset of pixels, but ignoring that for now for simplicity
for (i = 0; i < pixels_size; i++) {
  pixels[i] = data->color;
}

// composite tick()
for (i = 0; i < data->children_size; i++) {
  animation_tick(data->children[i], elapsed_time, pixels, pixels_size);
}

// rgb wave tick()
data->time_elapsed += elapsed_time;
for (i = 0; i < data->animation_pixels_size; i++) {
  int pixel_idx = data->animation_pixels[i];
  pixels[pixel_idx] = calculate_wave_color(pixel_idx, data->time_elapsed);
}
```

The two wave animations would each have their own `data->time_elapsed` state variables, but they would be reset to 0 when they are activated at the same time, and they would increment by the same value every tick, keeping them always in sync.